Implementation of GetCapabilities (list/XML) + Creation of a welcome page

guiMande opened the pull request:
Changes of the IndexController code, this class allows to display the index page which contains a list of all the processes of OrbisWPS, and a link to the xml file, reachable with HTTP request too.
Adds of two Controllers, WelcomeController which allows to display the welcome page, and GetCapabilitiesController which allows to display the xml page with the correct HTTP request, basics tests have been done. #2
Then, add of a basic HTML welcome page. #7
Finally, a beginnning of implementation of DescribeProcess have been done in IndexController for the moment. #3

guiMande committed:
Implements GetCapabilities

When the http request GetCapabilitites is called, a corresponding xml
file is returned. #3

guiMande committed:
Display the list of processes on index page

The index page display the list of available processes of OrbisWPS and
has a link to the xml file corresponding to the GetCapabilities result.

guiMande committed:
Re-organization of the code

Creation of controllers for each requests and creation of a very simple
welcome html page.
Beginnig of the implementation of DescribeProcess request.

ebocher commented on core-server/src/main/java/org/orbisgis/orbisserver/IndexController.java:
```diff
@@ -103,40 +111,41 @@ public Result index(@Parameter("service") String service, @Parameter("version")
+                    return ok(wpsCapabilitiesType);
                   }else{
-                    return ok("Error1");
+                    return badRequest("The request was not properly written");
```
Please use I18 properties to translate the message

ebocher commented on core-server/src/main/java/org/orbisgis/orbisserver/IndexController.java:
```diff
@@ -103,40 +111,41 @@ public Result index(@Parameter("service") String service, @Parameter("version")
      * Dirty method to display in the logger the list of the processes.
      * @throws JAXBException JAXB Exception.
      */
     private void simpleWpsRequest() throws JAXBException {
```
Why call this method simpleWpsRequest ?

ebocher commented on core-server/src/main/java/org/orbisgis/orbisserver/IndexController.java:
```diff
@@ -79,17 +81,23 @@
     private WpsServer wpsServer = WpsServerManager.getWpsServer();
 
-    private ArrayList<String> test = new ArrayList<String>();
+    private String test = "";
```
Don't understand the test variable.

ebocher reviewed (changes requested):
This PR needs some review. We need a clear architecture to organize the code, clear names for methods and also unit tests

SPalominos commented:
The travis failing is fixes, there was no .travis.yml file.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/GetCapabilitiesController.java:
```diff
@@ -0,0 +1,111 @@
+             if(version.equals("2.0.0")){
+               if(request!= null && !request.isEmpty()){
+                 if(request.equals("GetCapabilities")){
+                   try { indexController.simpleWpsRequest();  } catch (JAXBException ignored) {}
```
Please avoid to ignore an exception. At least log en error message.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/GetCapabilitiesController.java:
```diff
@@ -0,0 +1,111 @@
+                   try { indexController.simpleWpsRequest();  } catch (JAXBException ignored) {}
+                   return ok(indexController.wpsCapabilitiesType);
+                 }else{
+                   return badRequest("The request was not properly written");
```
For the request error message, you should give a look to the [wps standard](http://docs.opengeospatial.org/is/14-065/14-065.html#57), the exceptions are listed.

guiMande replied:
I have a doubt about how to use exceptions listed in the wps standard, should i do like in orbiswps with an ExceptionType ?

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/IndexController.java:
```diff
@@ -61,46 +66,62 @@
         //Simple example of getting information from the WpsServer
-        try { simpleWpsRequest(); } catch (JAXBException ignored) {}
-        return ok(render(index, "index", "Welcome"));
+        try { simpleWpsRequest();  } catch (JAXBException ignored) {}
```
Avoid to ignore an exception. At least log a message.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/IndexController.java:
```diff
@@ -127,10 +148,52 @@ private void simpleWpsRequest() throws JAXBException {
         List<ProcessSummaryType> list = wpsCapabilitiesType.getContents().getProcessSummary();
         for(ProcessSummaryType processSummaryType : list){
-            System.out.println(processSummaryType.getTitle().get(0).getValue());
+            //System.out.println(processSummaryType.getTitle().get(0).getValue());
```
If those lines are unused, just remove it.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/IndexController.java:
```diff
@@ -127,10 +148,52 @@ private void simpleWpsRequest() throws JAXBException {
+            //this.codeTypeList.add(processSummaryType.getIdentifier());
         }
+        this.list = list;
+        System.out.println(this.test);
```
Do not use `System.out.println()` if it is not absolutely necessary.

SPalominos commented on core-server/src/test/java/org/orbisgis/orbisserver/UnitTest.java:
```diff
@@ -43,14 +43,15 @@
  */
-public class UnitTest {
+public class UnitTest extends WisdomUnitTest {
 
```
You should complete the unit test class to test your changes.

guiMande committed:
Construction of an architecture+adds of some tests

The code has a better architecture now, and some tests have been adds to
BlackBoxIT, FluentLeniumIT, InContainer and UnitTest.
It now uses Logger and I18N to display and translate error messages.
It misses to use exceptions for GetCapabilities method
(InvalidParameterValue and MissingParameterValue).

guiMande committed:
Merge remote-tracking branch 'refs/remotes/orbisgis/master'

guiMande committed:
Update from orbisgis/master

Update from orbisgis/master

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/web/IndexController.java:
```diff
@@ -61,46 +71,67 @@
      */
     @View("index")
-    Template index;
+    public Template index;
```
Question, is it necessary to make this field public?

guiMande replied:
It is necessary if the UnitTest and the controller are not in the same package, two solutions for that : 
make it public or create UnitTest classes for each controllers and reproduce the same package.

SPalominos replied:
It's a good practice to have the tests in the same packages (but in the `test` folder) as the class tested. 
I.E. : if you have `/src/main/java/org/orbisgis/MyClass.java`, you should have `/src/main/test/java/org/orbisgis/MyClassTest.java`

For the access to the template for the test, maybe you could make it private and add a getter on that field.

guiMande replied:
I will change that and commit it later.

SPalominos replied:
Ok, great!

guiMande replied:
I have all units Tests in the right place now and IT Tests location does not change, they are integration tests which check if the returned page after a http request has the correct content, so there is no specific package for them.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/xml/GetCapabilitiesController.java:
```diff
@@ -0,0 +1,127 @@
+    *
+    * @return the xml file
+    */
+   @Route(method = HttpMethod.GET, uri = "/orbisserver/ows")
```
Replace `ows` by `wps`

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/xml/GetCapabilitiesController.java:
```diff
@@ -0,0 +1,127 @@
+    * HTTP GET request on the "/orbisserv/ows" URL. 
+    * A good request should be http://localhost:9000/orbisserver/ows?service=WPS&version=2.0.0&request=GetCapabilities
+    *
+    * @Parameter the service
```
The javadoc parameter tag work this way : `@param theParameterName The description of the parameter.`

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/xml/GetCapabilitiesController.java:
```diff
@@ -0,0 +1,127 @@
+                 if(request.equals("GetCapabilities")){
+                   try {
+                     indexController.GetXMLFromGetCapabilities();
+                   } catch (JAXBException ignored) {
```
If the exception is not ignored, change it name.

SPalominos commented:
Great @guiMande, just few things to fix. I will take a look on the Travis building to find why it has failed.

guiMande committed:
Adds of some changes + port 8080

The server is now using the port 8080.
Creation of packages for tests classes, changes of the javadoc, and some
others changes.

ebocher commented on core-server/src/test/java/org/orbisgis/orbisserver/control/web/IndexControllerTest.java:
```diff
@@ -0,0 +1,73 @@
+ *
+ * OrbisServer is distributed under LGPL 3 license.
+ *
+ * Copyright (C) 2007-2014 CNRS (IRSTV FR CNRS 2488)
```
No Copyright to IRSTV

guiMande committed:
Adds of Exceptions for GetCapabilities + some changes

Display an InvalidParameterValue exception or a MissingParameterValue
exception in the terminal if the request is not well written

ebocher commented on core-server/src/main/configuration/application.conf:
```diff
@@ -39,6 +39,11 @@ monitor {
 
+http.port = 8080
+https.port = 8080
+vertx.http.port = 8080
```
Please add doc on the readme file to explain this config file

ebocher commented on core-server/src/main/java/org/orbisgis/orbisserver/control/xml/GetCapabilitiesController.java:
```diff
@@ -79,18 +79,20 @@
     * @return the xml file
     */
-   @Route(method = HttpMethod.GET, uri = "/orbisserver/ows")
+   @Route(method = HttpMethod.GET, uri = "/orbisserver/wps")
```
Not sure about this URI. Why remove ows ?

SPalominos replied:
It was me which requested this change. I just figure out that it is better to keep `ows` instead of `wps`, my bad. @guiMande sorry for the change, could you put back `ows` ?

ebocher commented on core-server/src/main/java/org/orbisgis/orbisserver/control/xml/GetCapabilitiesController.java:
```diff
@@ -101,8 +103,8 @@ public Result displayXML(@Parameter("service") String service, @Parameter("versi
-                   } catch (JAXBException ignored) {
-                     LOGGER.error(I18N.tr("Unable to parse the incoming xml. \nCause : {0}.", ignored.toString()));
+                   } catch (JAXBException e) {
+                     LOGGER.error(I18N.tr("Unable to parse the incoming xml. \nCause : {0}.", e.getMessage()));
```
Is this the correct response when a GetXMLFromGetCapabilities fails ? How the message is returned to the client ?

guiMande replied:
I change the error message to "Unable to get the xml file corresponding to the GetCapabilities request."
I create a html error page with the exception and the text of the excpetion which is returned when the request fails, is that ok ?

SPalominos replied:
In the OGC standards, exceptions are defined. In the case of GetCapabilities, the exception in the WPS standard is definec [here](http://docs.opengeospatial.org/is/14-065/14-065.html#53), which send you to the document [OGC 06-121r9](https://www.google.fr/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&ved=0ahUKEwiFqYeDi_TTAhXLXhoKHUZlDYYQFggjMAA&url=https%3A%2F%2Fportal.opengeospatial.org%2Ffiles%2F%3Fartifact_id%3D38867&usg=AFQjCNHmxITnXCBqUrmqS7VsPxS6l7mPHQ&sig2=8gpBLhVKjjMcVyckiNq0HQ&cad=rja), section 7.4.

SPalominos replied:
The exception should be return in the same way as the xml content to allow the client to receive the exception.

ebocher commented on core-server/src/main/java/org/orbisgis/orbisserver/control/xml/GetCapabilitiesController.java:
```diff
@@ -101,8 +103,8 @@ public Result displayXML(@Parameter("service") String service, @Parameter("versi
                  if(request.equals("GetCapabilities")){
                    try {
                      indexController.GetXMLFromGetCapabilities();
```
rename to getXMLFromGetCapabilities

ebocher commented on core-server/src/test/java/org/orbisgis/orbisserver/control/xml/GetCapabilitiesControllerTest.java:
```diff
@@ -37,55 +37,21 @@
+ /**
+  * A couple of unit tests.
+  */
+  public class GetCapabilitiesControllerTest extends WisdomUnitTest {
```
testGetCapabilities must be improved with more use cases.

guiMande committed:
Adds of tests for GetCapabilities + some changes

More tests of GetCapabilities have been add, README file up to date, wps
change to ows, method title up to date, exceptions are now returning xml
file corresponding to the exception  to the client .

ebocher commented on core-server/src/main/java/org/orbisgis/orbisserver/control/xml/GetCapabilitiesController.java:
```diff
@@ -69,44 +69,48 @@
+     isFailed = false;
 
-     //Simple example of getting information from the WpsServer
+     //Simple example of getting informaaation from the WpsServer
```
information ?

guiMande replied:
My bad, i'll change this

guiMande committed:
Javadoc Correction

Change of informaaation to information

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/web/IndexController.java:
```diff
@@ -61,21 +70,34 @@
+    /** Gets the instance of the WpsServer. */
     private WpsServer wpsServer = WpsServerManager.getWpsServer();
+    /** List all of the OrbisWPS processes into a String which be displayed on the index page. */
+    private String processesList = "";
```
Why `processesList` is not a local variable as it is used only in the `index()` method ?

guiMande replied:
Should i change getXMLFromGetCapabilities to return the processList ?

guiMande replied:
My bad, i didn't read all the comments.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/web/IndexController.java:
```diff
@@ -61,21 +70,34 @@
+    /** List all of the OrbisWPS processes into a String which be displayed on the index page. */
+    private String processesList = "";
+    /**  Used to display the corresponding xml file to GetCapabilities method. */
+    public WPSCapabilitiesType wpsCapabilitiesType = null;
```
Do not make a variable be `public` unless it is a `static` one. If you need to make it accessible, use getters and setters.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/web/IndexController.java:
```diff
@@ -61,21 +70,34 @@
+    /**  Used to display the corresponding xml file to GetCapabilities method. */
+    public WPSCapabilitiesType wpsCapabilitiesType = null;
+    /** Used by methods like DescribeProcessRequest to get all the ProcessSummaryType */
+    public List<ProcessSummaryType> processSummaryTypeList = null;
```
Same thing about `public` visibility.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/web/IndexController.java:
```diff
@@ -84,23 +106,29 @@
      * @throws JAXBException JAXB Exception.
      */
-    private void simpleWpsRequest() throws JAXBException {
+    public void getXMLFromGetCapabilities() throws JAXBException {
```
Maybe this method should return either the `WpsCapabilitiesType` object or the process list instead of setting variables. (Linked to the comment about local variable)

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/xml/GetCapabilitiesController.java:
```diff
@@ -0,0 +1,167 @@
+  /**
+    * IndexController's object, used to get some useful attributes.
+    */
+   private IndexController indexController = new IndexController();
```
You should not instantiate an `IndexController` inside the `GetCapabilitiesController` if it is only to get the `WpsCapabilitiesType` object. If you need to manage the access to the wps information and to access them from different classes, maybe the better thing to do is to have one class which return the well formatted information (xml or process list).

guiMande replied:
That is the same thing if I create a new class that instantiate an `IndexController` and has some getters and setters, which will be used by `GetCapabilitiesController` no ?

SPalominos commented on core-server/src/test/java/org/orbisgis/orbisserver/control/xml/GetCapabilitiesControllerTest.java:
```diff
@@ -0,0 +1,152 @@
+          }
+        }).invoke();
+
+        if(!controller.isFailed()){
```
In a test you should verify a desired behaviour, so you should know if your controller has failed or not.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/web/IndexController.java:
```diff
@@ -61,21 +70,34 @@
+    /** Used by methods like DescribeProcessRequest to get all the ProcessSummaryType */
+    public List<ProcessSummaryType> processSummaryTypeList = null;
+    /** List of all identifier of all the processes */
+    private List<CodeType> codeTypeList = null;
```
`codeTypeList` seems to be never used.

SPalominos reviewed (changes requested):
Be careful, your tabulation are inconsistent (sometimes 2 space, sometimes 4). Please use 4 space tabulations.

guiMande committed:
Public parameters removed + Modification of WpsServerManager

Public parameters and unused parameters have been removed or change to
local variables. Add of WpsCapabilities getter and
getXMLFromGetCapabilities returns the processes list into a String.
Those methods are now in WpsServerManager for easier access.
Test of failed GetCapabilititesController removed because it's already
done in OrbisWps.
Tabulation corrected.
Licence update. #13

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/web/IndexController.java:
```diff
@@ -0,0 +1,98 @@
+     * Injects a template named 'index'.
+     */
+    @View("index")
+    Template index;
```
Can the field `index` be `private` instead of `package` ?

guiMande replied:
It can be private, this only affect the test, using Action like testGetCapabilitiesController instead of mock.

SPalominos replied:
Ok, so keep it this way (package) for now.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/web/WelcomeController.java:
```diff
@@ -0,0 +1,70 @@
+     * Injects a template named 'welcome'.
+     */
+    @View("welcome")
+    Template welcome;
```
Can the field `index` be `private` instead of `package` ?

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/web/IndexController.java:
```diff
@@ -0,0 +1,98 @@
+import javax.xml.bind.JAXBException;
+
+/**
+ * Instance of DefaultController containing a WpsServer.
```
Maybe you could make a more accurate description of the class.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/web/IndexController.java:
```diff
@@ -0,0 +1,98 @@
+ */
+@Controller
+public class IndexController extends DefaultController {
+    /** Gets the instance of the WpsServer. */
```
It give the instance of an `WpsServerManager` and not an instance of `WpsServer`.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/web/IndexController.java:
```diff
@@ -0,0 +1,98 @@
+    public Result index() {
+        /** List all of the OrbisWPS processes into a String which will be displayed on the index page. */
+        String resultList = "";
+        //Simple example of getting information from the WpsServer
```
This comment seem to be nonsense.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/web/IndexController.java:
```diff
@@ -0,0 +1,98 @@
+            LOGGER.error(I18N.tr("Unable to get the xml file corresponding to the GetCapabilities request. \nCause : {0}.", e.getMessage()));
+            resultList = "Unable to get the xml file";
+        }
+        return ok(render(index, "liste", resultList));
```
You should replace `liste` by `list` to keep all the code in English.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/xml/GetCapabilitiesController.java:
```diff
@@ -0,0 +1,144 @@
+import javax.xml.bind.JAXBException;
+
+/**
+ * Instance of DefaultController containing a WpsServer.
```
Maybe you could make a more accurate description of the class.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/control/xml/GetCapabilitiesController.java:
```diff
@@ -0,0 +1,144 @@
+ */
+@Controller
+public class GetCapabilitiesController extends DefaultController {
+    /** IndexController's object, used to get some useful attributes */
```
This description seems to be nonsense.

SPalominos commented on core-server/src/main/java/org/orbisgis/orbisserver/manager/WpsServerManager.java:
```diff
@@ -121,16 +132,26 @@ private void simpleWpsRequest() throws JAXBException {
+     * @Return the wpsCapabilitiesType object.
+     */
+    public WPSCapabilitiesType getWPSCapabilitiesType() {
+        return this.wpsCapabilitiesType;
```
You should ask the `WpsCapabilitiesType` to the `WpsServer` each time because it can changes (processes can be add, removed ...) like you are doing in the `getXMLFromGetCapabilities()` method.

guiMande replied:
Should i add `getXMLFromGetCapabilities();` before the return or rewrite a part of this method in getWPSCapabilitiesType() ?

SPalominos replied:
The reverse. You should put the request to the WPS server in the `getWPSCapabilitiesType()` method. And in the `getXMLFromGetCapabilities()` methods you should use the `WPSCapabilitiesType` Object get from `getWPSCapabilitiesType()`.

guiMande replied:
So the `GetXMLFromGetCapabilities()` method returns the string list, and might be rename, and the `GetWPSCapabilitiesType()` method can return the WPSCapabilitiesType Object, is that ok ?

SPalominos commented on core-server/src/main/resources/templates/index.thl.html:
```diff
@@ -49,21 +49,24 @@
 
 </div>
 
+<pre th:text="${liste}">Welcome!</pre>
```
You should replace `liste` by `list` to keep all the code in English.

SPalominos reviewed (changes requested):
Great, you are almost done. Some little very changes (mostly javadoc). I will perform the test and I will make a feed back.

SPalominos commented on core-server/src/test/java/org/orbisgis/orbisserver/control/xml/GetCapabilitiesControllerTest.java:
```diff
@@ -0,0 +1,140 @@
+          }).invoke();
+
+          assertThat(status(result)).isEqualTo(OK);
+          assertThat(toString(result)).contains("net.opengis.wps._2_0.WPSCapabilitiesType@");
```
In your tests your are using Assertj library. In the OrbisGIS projects we use JUnitso you should change your assertion to use the class `org.junit.Assert` class.
In this case, your test should be :
``` java
Assert.assertTrue(toString(result).contains("net.opengis.wps._2_0.WPSCapabilitiesType@");

guiMande replied:
How do you transform `assertThat(find(".unitTest")).hasText(OrbisServer");` in FluentLeniumIT to a JUnit assert ?

SPalominos replied:
You have part of the answer just above.
`Assert.assertTrue(find(".unitTest").contains("OrbisServer"));`

SPalominos reviewed (changes requested):
You should changes you test assertions to use the JUnit library (org.junit.Assert class) instead of Assertj.

guiMande committed:
Javadoc changes + Assert up to date

Some Javadoc changes, liste renamed to list,
modification of WpsServerManager, now we have two methods
getXMLFromGetCapabilities and getListFromGetCapabilities.
Tests are now using Assert of JUnit instead of Assertj.

SPalominos commented:
Travis building error is due to java versions, I'll fixe that asap.

guiMande committed:
Changes in travis.yml

Openjdk 7 suppresion

SPalominos commented:
Great, travis fixed, PR merged.
